File: backend/internal/models/supply.go
package models

import (
	"time"

	"go.mongodb.org/mongo-driver/v2/bson"
)

type SupplyCategory string

const (
	SupplyFertilizer         SupplyCategory = "fertilizer"
	SupplyPesticideHerbicide SupplyCategory = "pesticide_herbicide_fungicide"
	SupplySeedsSeedlings     SupplyCategory = "seeds_seedlings"
	SupplyTools              SupplyCategory = "tools"
	SupplyPPE                SupplyCategory = "ppe"
)

type SupplyOrderStatus string

const (
	SupplyOrderPending    SupplyOrderStatus = "pending"
	SupplyOrderProcessing SupplyOrderStatus = "processing"
	SupplyOrderShipped    SupplyOrderStatus = "shipped_ready"
	SupplyOrderCompleted  SupplyOrderStatus = "completed"
	SupplyOrderCancelled  SupplyOrderStatus = "cancelled"
)

type DeliveryMethod string

const (
	DeliveryShip   DeliveryMethod = "delivery"
	DeliveryPickup DeliveryMethod = "pickup"
)

// PaymentMethod is the mode of payment chosen by the buyer at checkout.
type PaymentMethod string

const (
	// PaymentCOD — Cash on Delivery. Payment is collected by the supplier/rider upon delivery.
	PaymentCOD PaymentMethod = "cod"
	// PaymentGCash — GCash e-wallet (Philippines). Processed via PayMongo (future integration).
	PaymentGCash PaymentMethod = "gcash"
	// PaymentMaya — Maya (formerly PayMaya) e-wallet (Philippines). Via PayMongo (future).
	PaymentMaya PaymentMethod = "maya"
	// PaymentBankTransfer — Online bank transfer / InstaPay / PESONet. Via Dragonpay (future).
	PaymentBankTransfer PaymentMethod = "bank_transfer"
	// PaymentCard — Credit or debit card (Visa/Mastercard). Via PayMongo (future).
	PaymentCard PaymentMethod = "card"
)

// PaymentStatus tracks the lifecycle of the payment for an order.
type PaymentStatus string

const (
	// PaymentStatusPending — Payment not yet received (default for COD and newly placed orders).
	PaymentStatusPending PaymentStatus = "pending_payment"
	// PaymentStatusPaid — Payment confirmed (COD: marked by supplier; online: via gateway webhook).
	PaymentStatusPaid PaymentStatus = "paid"
	// PaymentStatusFailed — Online payment attempt failed or was rejected.
	PaymentStatusFailed PaymentStatus = "failed"
	// PaymentStatusRefunded — Payment was refunded to the buyer.
	PaymentStatusRefunded PaymentStatus = "refunded"
)

// SupplyProduct represents an agricultural input item listed by a supplier.
type SupplyProduct struct {
	ID            bson.ObjectID  `bson:"_id,omitempty"       json:"id"`
	SupplierID    bson.ObjectID  `bson:"supplier_id"         json:"supplierId"`
	SupplierName  string         `bson:"supplier_name"       json:"supplierName"`
	IsVerified    bool           `bson:"is_verified"         json:"isVerified"` // Supplier accreditation flag
	Name          string         `bson:"name"                json:"name"`
	Category      SupplyCategory `bson:"category"            json:"category"`
	Description   string         `bson:"description"         json:"description"`
	Price         float64        `bson:"price"               json:"price"`
	StockQuantity int            `bson:"stock_quantity"      json:"stockQuantity"`
	Unit          string         `bson:"unit"                json:"unit"` // e.g. "bag", "liter", "pack", "piece"
	Images        []string       `bson:"images,omitempty"    json:"images,omitempty"`
	CreatedAt     time.Time      `bson:"created_at"          json:"createdAt"`
	UpdatedAt     time.Time      `bson:"updated_at"          json:"updatedAt"`
}

// CreateSupplyProductRequest is the JSON payload to list a supply product.
type CreateSupplyProductRequest struct {
	Name          string         `json:"name"`
	Category      SupplyCategory `json:"category"`
	Description   string         `json:"description"`
	Price         float64        `json:"price"`
	StockQuantity int            `json:"stockQuantity"`
	Unit          string         `json:"unit"`
	Images        []string       `json:"images,omitempty"`
}

// UpdateSupplyProductRequest is the JSON payload to update a product.
type UpdateSupplyProductRequest struct {
	Name          *string         `json:"name,omitempty"`
	Category      *SupplyCategory `json:"category,omitempty"`
	Description   *string         `json:"description,omitempty"`
	Price         *float64        `json:"price,omitempty"`
	StockQuantity *int            `json:"stockQuantity,omitempty"`
	Unit          *string         `json:"unit,omitempty"`
	Images        []string        `json:"images,omitempty"`
}

// SupplyOrderItem represents an item line inside a supply order.
type SupplyOrderItem struct {
	ProductID    bson.ObjectID `bson:"product_id"    json:"productId"`
	ProductName  string        `bson:"product_name"  json:"productName"`
	ProductImage string        `bson:"product_image,omitempty" json:"productImage,omitempty"`
	Quantity     int           `bson:"quantity"      json:"quantity"`
	PricePerItem float64       `bson:"price_per_item" json:"pricePerItem"`
}

// SupplyOrder represents an order placed by a farmer to a supplier.
type SupplyOrder struct {
	ID              bson.ObjectID     `bson:"_id,omitempty"        json:"id"`
	BuyerID         bson.ObjectID     `bson:"buyer_id"             json:"buyerId"`
	BuyerName       string            `bson:"buyer_name"           json:"buyerName"`
	SupplierID      bson.ObjectID     `bson:"supplier_id"          json:"supplierId"`
	SupplierName    string            `bson:"supplier_name"        json:"supplierName"`
	Items           []SupplyOrderItem `bson:"items"                json:"items"`
	TotalAmount     float64           `bson:"total_amount"         json:"totalAmount"`
	DeliveryMethod  DeliveryMethod    `bson:"delivery_method"      json:"deliveryMethod"`
	DeliveryAddress string            `bson:"delivery_address,omitempty" json:"deliveryAddress,omitempty"`
	Status          SupplyOrderStatus `bson:"status"               json:"status"`
	// Payment fields
	PaymentMethod   PaymentMethod     `bson:"payment_method"       json:"paymentMethod"`
	PaymentStatus   PaymentStatus     `bson:"payment_status"       json:"paymentStatus"`
	PaymentNote     string            `bson:"payment_note,omitempty" json:"paymentNote,omitempty"` // e.g. reference number for online payments
	CreatedAt       time.Time         `bson:"created_at"           json:"createdAt"`
	UpdatedAt       time.Time         `bson:"updated_at"           json:"updatedAt"`
}

// CreateOrderItemRequest represents an item entry during checkout.
type CreateOrderItemRequest struct {
	ProductID string `json:"productId"`
	Quantity  int    `json:"quantity"`
}

// CreateSupplyOrderRequest is the JSON body for checkout.
type CreateSupplyOrderRequest struct {
	Items           []CreateOrderItemRequest `json:"items"`
	DeliveryMethod  DeliveryMethod           `json:"deliveryMethod"`
	DeliveryAddress string                   `json:"deliveryAddress,omitempty"`
	// PaymentMethod defaults to "cod" if omitted.
	PaymentMethod   PaymentMethod            `json:"paymentMethod"`
}

// UpdateSupplyOrderStatusRequest updates an order's fulfillment status.
type UpdateSupplyOrderStatusRequest struct {
	Status SupplyOrderStatus `json:"status"`
}

// UpdatePaymentStatusRequest lets a supplier confirm COD payment receipt
// or lets the system update online payment status after gateway webhook.
type UpdatePaymentStatusRequest struct {
	PaymentStatus PaymentStatus `json:"paymentStatus"`
	PaymentNote   string        `json:"paymentNote,omitempty"`
}

File: web/src/types/supply.ts
export type SupplyCategory =
  | 'fertilizer'
  | 'pesticide_herbicide_fungicide'
  | 'seeds_seedlings'
  | 'animal_feeds'
  | 'vet_medicines'
  | 'irrigation'
  | 'machinery_equipment'
  | 'nursery_greenhouse'
  | 'packaging_storage'
  | 'tools'
  | 'ppe';

export interface SupplyCategoryConfig {
  key: SupplyCategory;
  label: string;
  icon: string;
  suggestedUnits: string[];
  placeholderName: string;
}

export const SUPPLY_CATEGORIES: SupplyCategoryConfig[] = [
  {
    key: 'fertilizer',
    label: 'Fertilizers & Soil Nutrients',
    icon: '🌱',
    suggestedUnits: ['sack (50kg)', 'bag (25kg)', 'kg', 'bottle (1L)', 'liter'],
    placeholderName: 'e.g. Complete 14-14-14, Urea 46-0-0, Organic Vermicast',
  },
  {
    key: 'pesticide_herbicide_fungicide',
    label: 'Crop Protection & Chemicals',
    icon: '🧪',
    suggestedUnits: ['bottle (1L)', 'bottle (500ml)', 'sachet (100g)', 'kg', 'pack'],
    placeholderName: 'e.g. Broad-spectrum Insecticide, Glyphosate Herbicide, Mancozeb Fungicide',
  },
  {
    key: 'seeds_seedlings',
    label: 'Seeds & Planting Materials',
    icon: '🌽',
    suggestedUnits: ['can (50g)', 'can (100g)', 'pouch', 'pack (1kg)', 'seedling tray', 'piece'],
    placeholderName: 'e.g. Hybrid Yellow Corn Seeds, Diamante Max Tomato Seeds, Grafted Mango Seedlings',
  },
  {
    key: 'animal_feeds',
    label: 'Animal Feeds & Nutrition',
    icon: '🌾',
    suggestedUnits: ['sack (50kg)', 'bag (25kg)', 'kg', 'bundle', 'pack'],
    placeholderName: 'e.g. Chick Booster Crumble, Hog Grower Pellets, Cattle Concentrate, Rice Bran (Darak)',
  },
  {
    key: 'vet_medicines',
    label: 'Veterinary Medicine & Biologics',
    icon: '💉',
    suggestedUnits: ['bottle (100ml)', 'bottle (1L)', 'vial', 'sachet (10g)', 'sachet (100g)', 'box'],
    placeholderName: 'e.g. Vitstress Electrolytes + Vitamin B-Complex, Albendazole Dewormer, Amoxicillin Powder',
  },
  {
    key: 'irrigation',
    label: 'Irrigation & Water Systems',
    icon: '💧',
    suggestedUnits: ['roll (100m)', 'roll (500m)', 'piece', 'set', 'meter'],
    placeholderName: 'e.g. Drip Irrigation Tape Kit 16mm, PE Pipe, Micro Sprinkler Set, Automatic Hog Nipple Drinker',
  },
  {
    key: 'machinery_equipment',
    label: 'Farm Machinery & Power Equipment',
    icon: '⚙️',
    suggestedUnits: ['unit', 'set', 'piece'],
    placeholderName: 'e.g. 2-in-1 Rechargeable Battery Knapsack Sprayer 16L, 4-Stroke Brush Cutter, 3-inch Water Pump',
  },
  {
    key: 'nursery_greenhouse',
    label: 'Greenhouse, Mulch & Nursery',
    icon: '🏡',
    suggestedUnits: ['roll', 'piece (tray)', 'bundle (100pcs)', 'meter'],
    placeholderName: 'e.g. Silver-Black Plastic Mulch 1.2m x 400m, Black Shade Net 70%, 104-Hole Seedling Trays',
  },
  {
    key: 'packaging_storage',
    label: 'Packaging, Sacks & Crates',
    icon: '📦',
    suggestedUnits: ['piece', 'bundle (50pcs)', 'bundle (100pcs)', 'bale'],
    placeholderName: 'e.g. 50kg Woven Polypropylene Sacks, Heavy-Duty Plastic Harvest Crates (Kaing)',
  },
  {
    key: 'tools',
    label: 'Hand Tools & Implements',
    icon: '🔧',
    suggestedUnits: ['piece', 'set', 'pair'],
    placeholderName: 'e.g. Heavy-Duty Carabao Bolo, Digging Shovel, Pruning Shears, Garden Rake',
  },
  {
    key: 'ppe',
    label: 'Safety Gear & PPE',
    icon: '🥽',
    suggestedUnits: ['pair', 'piece', 'set', 'box'],
    placeholderName: 'e.g. Chemical Spraying Respirator Mask, Rubber Knee Boots, Heavy-Duty Nitrile Gloves',
  },
];

export type DeliveryMethod = 'delivery' | 'pickup';

export type SupplyOrderStatus =
  | 'pending'
  | 'quoted'
  | 'processing'
  | 'shipped_ready'
  | 'completed'
  | 'cancelled';

/**
 * PaymentMethod — mirrors backend models.PaymentMethod constants.
 * - cod: Cash on Delivery (no gateway required, supplier confirms on delivery)
 * - gcash: GCash e-wallet via PayMongo (future)
 * - maya: Maya (PayMaya) e-wallet via PayMongo (future)
 * - bank_transfer: InstaPay / PESONet via Dragonpay (future)
 * - card: Visa / Mastercard via PayMongo (future)
 */
export type PaymentMethod = 'cod' | 'gcash' | 'maya' | 'bank_transfer' | 'card';

/**
 * PaymentStatus — mirrors backend models.PaymentStatus constants.
 * - pending_payment: default for all new orders; COD awaits delivery confirmation
 * - paid: confirmed paid (COD: by supplier; online: by gateway webhook)
 * - failed: online payment attempt was declined
 * - refunded: payment was reversed
 */
export type PaymentStatus = 'pending_payment' | 'paid' | 'failed' | 'refunded';

export interface SupplyProduct {
  id: string;
  supplierId: string;
  supplierName: string;
  isVerified: boolean;
  name: string;
  category: SupplyCategory;
  description: string;
  price: number;
  stockQuantity: number;
  unit: string;
  location?: string;
  images?: string[];
  createdAt: string;
  updatedAt: string;
}

export interface CreateSupplyProductPayload {
  name: string;
  category: SupplyCategory;
  description: string;
  price: number;
  stockQuantity: number;
  unit: string;
  location?: string;
  images?: string[];
}

export interface SupplyOrderItem {
  productId: string;
  productName: string;
  productImage?: string;
  quantity: number;
  pricePerItem: number;
}

export interface SupplyOrder {
  id: string;
  buyerId: string;
  buyerName: string;
  supplierId: string;
  supplierName: string;
  items: SupplyOrderItem[];
  subtotal?: number;
  shippingFee?: number;
  totalAmount: number;
  deliveryMethod: DeliveryMethod;
  deliveryAddress?: string;
  status: SupplyOrderStatus;
  paymentMethod: PaymentMethod;
  paymentStatus: PaymentStatus;
  paymentNote?: string;
  createdAt: string;
  updatedAt: string;
}

export interface CreateSupplyOrderPayload {
  items: { productId: string; quantity: number }[];
  deliveryMethod: DeliveryMethod;
  deliveryAddress?: string;
  /** Defaults to 'cod' on the backend if omitted. */
  paymentMethod: PaymentMethod;
}

export interface UpdatePaymentStatusPayload {
  paymentStatus: PaymentStatus;
  paymentNote?: string;
}
